### PR TITLE
new(pkg/driver): try to fetch kernel headers leveraging driverkit library when building drivers

### PR DESCRIPTION
**Pull request by FedeDP**
**What type of PR is this?**

/kind feature

**Any specific area of the project related to this PR?**

/area library

**What this PR does / why we need it**:

This PR introduces a mechanism to leverage driverkit library to automatically download kernel headers for driverkit supported distros, when building drivers.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

### CODE REVIEW

File: pkg/driver/distro/distro.go [hunk 1]
```diff
@@ -127,6 +126,46 @@ func getOSReleaseDistro(kr *kernelrelease.KernelRelease) (Distro, error) {
...
+	if err != nil {
+		return nil, err
+	}
+	if len(kernelheaders) < minimumURLs {
+		return nil, fmt.Errorf("not enough headers packages found; expected %d, found %d", minimumURLs, len(kernelheaders))
+	}
+	return kernelheaders, nil
+}
+
+func downloadKernelHeaders(headers []string) (string, error) {
```
Comment (FedeDP) [inline pkg/driver/distro/distro.go:160]: Here we will need to leverage new code in https://github.com/falcosecurity/driverkit/pull/324 to be able to run the templated scripts per distro (driverkit builder) to extract downloaded kernel sources and take the KERNELDIR path.

**Comment by FedeDP** [comment]
Output for a bpf probe build against drivers `6.0.1+driver` on my archlinux machine using downloaded headers:
```
2024-03-14 09:00:59 INFO  Running falcoctl driver install
                      ├ driver version: 6.0.1+driver
                      ├ driver type: ebpf
                      ├ driver name: falco
                      ├ compile: true
                      ├ download: true
                      ├ arch: x86_64
                      ├ kernel release: 6.8.0-arch1-1
                      └ kernel version: #1 SMP PREEMPT_DYNAMIC Mon, 11 Mar 2024 03:12:55 +0000
2024-03-14 09:00:59 INFO  Found distro target: arch
2024-03-14 09:00:59 INFO  Removing eBPF probe symlink path: /root/.falco/falco-bpf.o
2024-03-14 09:00:59 INFO  Trying to download a driver. url: https://download.falco.org/driver/6.0.1%2Bdriver/x86_64/falco_arch_6.8.0-arch1-1_1.o
2024-03-14 09:01:00 WARN  Non-200 response from url. code: 404
2024-03-14 09:01:00 WARN  unable to find a prebuilt driver
2024-03-14 09:01:07 INFO  Mounting debugfs for bpf driver.
make: ingresso nella directory «/usr/src/falco-6.0.1+driver/bpf»
make -C /tmp/kernel M=$PWD
make[1]: ingresso nella directory «/tmp/kernel»
clang -I./arch/x86/include -I./arch/x86/include/generated  -I./include -I./arch/x86/include/uapi -I./arch/x86/include/generated/uapi -I./include/uapi -I./include/generated/uapi -include ./include/linux/compiler-version.h -include ./include/linux/kconfig.h \
        -D__KERNEL__ -fmacro-prefix-map=./= \
         \
         \
        -D__KERNEL__ \
        -D__BPF_TRACING__ \
        -Wno-gnu-variable-sized-type-not-at-end \
        -Wno-address-of-packed-member \
        -fno-jump-tables \
        -fno-stack-protector \
        -Wno-tautological-compare \
        -Wno-unknown-attributes \
        -O2 -g -emit-llvm -c /usr/src/falco-6.0.1+driver/bpf/probe.c -o /usr/src/falco-6.0.1+driver/bpf/probe.ll
In file included from /usr/src/falco-6.0.1+driver/bpf/probe.c:16:
In file included from ./include/linux/sched.h:12:
In file included from ./arch/x86/include/asm/current.h:10:
In file included from ./include/linux/cache.h:6:
In file included from ./arch/x86/include/asm/cache.h:5:
In file included from ./include/linux/linkage.h:8:
In file included from ./arch/x86/include/asm/linkage.h:6:
./arch/x86/include/asm/ibt.h:77:8: warning: 'nocf_check' attribute ignored; use -fcf-protection to enable the attribute [-Wignored-attributes]
   77 | extern __noendbr u64 ibt_save(bool disable);
      |        ^
./arch/x86/include/asm/ibt.h:32:34: note: expanded from macro '__noendbr'
   32 | #define __noendbr       __attribute__((nocf_check))
      |                                        ^
./arch/x86/include/asm/ibt.h:78:8: warning: 'nocf_check' attribute ignored; use -fcf-protection to enable the attribute [-Wignored-attributes]
   78 | extern __noendbr void ibt_restore(u64 save);
      |        ^
./arch/x86/include/asm/ibt.h:32:34: note: expanded from macro '__noendbr'
   32 | #define __noendbr       __attribute__((nocf_check))
      |                                        ^
In file included from /usr/src/falco-6.0.1+driver/bpf/probe.c:26:
/usr/src/falco-6.0.1+driver/bpf/fillers.h:2851:22: error: no member named 'i_mtime' in 'struct inode'; did you mean '__i_mtime'?
 2851 |         time = _READ(inode->i_mtime);
      |                             ^~~~~~~
      |                             __i_mtime
/usr/src/falco-6.0.1+driver/bpf/plumbing_helpers.h:23:28: note: expanded from macro '_READ'
   23 | #define _READ(P) ({ typeof(P) _val;                                     \
      |                            ^
./include/linux/fs.h:677:20: note: '__i_mtime' declared here
  677 |         struct timespec64       __i_mtime;
      |                                 ^
In file included from /usr/src/falco-6.0.1+driver/bpf/probe.c:26:
/usr/src/falco-6.0.1+driver/bpf/fillers.h:2851:22: error: no member named 'i_mtime' in 'struct inode'; did you mean '__i_mtime'?
 2851 |         time = _READ(inode->i_mtime);
      |                             ^~~~~~~
      |                             __i_mtime
/usr/src/falco-6.0.1+driver/bpf/plumbing_helpers.h:24:51: note: expanded from macro '_READ'
   24 |                     bpf_probe_read_kernel(&_val, sizeof(_val), &P);     \
      |                                                                 ^
./include/linux/fs.h:677:20: note: '__i_mtime' declared here
  677 |         struct timespec64       __i_mtime;
      |                                 ^
2 warnings and 2 errors generated.
make[3]: *** [/usr/src/falco-6.0.1+driver/bpf/Makefile:54: /usr/src/falco-6.0.1+driver/bpf/probe.o] Error 1
make[2]: *** [/tmp/kernel/Makefile:1921: /usr/src/falco-6.0.1+driver/bpf] Error 2
make[1]: *** [Makefile:240: __sub-make] Error 2
make[1]: uscita dalla directory «/tmp/kernel»
make: *** [Makefile:38: all] Error 2
make: uscita dalla directory «/usr/src/falco-6.0.1+driver/bpf»
2024-03-14 09:01:08 ERROR failed: exit status 2
```
As you can see:
* we use `/tmp/kernel` as src directory during the build
* the build fails because driver 6.0.1+driver did not support kernel 6.8
* `/tmp/kernel` and `/tmp/kernel-download` folders are cleaned up upon leaving

**Comment by FedeDP** [comment]
Oras bump was moved to https://github.com/falcosecurity/falcoctl/pull/478. 
Since driverkit master uses new oras, the bump is needed.

**Comment by FedeDP** [comment]
Output for a kmod build (with dkms):
```
...
Kernel config /tmp/kernel/.config not found, modules won't be signed
DIRECTIVE: MAKE="'/tmp/falco-dkms-make'"
Creating symlink /var/lib/dkms/falco/6.0.1+driver/source -> /usr/src/falco-6.0.1+driver

Building module:
make -C /tmp/kernel M=/var/lib/dkms/falco/6.0.1+driver/build clean
make: ingresso nella directory «/tmp/kernel»
  CLEAN   /var/lib/dkms/falco/6.0.1+driver/build/Module.symvers
make: uscita dalla directory «/tmp/kernel»

{ '/tmp/falco-dkms-make'; } >> /var/lib/dkms/falco/6.0.1+driver/build/make.log 2>&1
(bad exit status: 2)
Error! Bad return status for module build on kernel: 6.8.0-arch1-1 (x86_64)
Consult /var/lib/dkms/falco/6.0.1+driver/build/make.log for more information.
2024-03-14 15:07:07 WARN  Running dkms build failed. Dumping dkms log. file: /var/lib/dkms/falco/6.0.1+driver/build/make.log
...
```

**Comment by FedeDP** [comment]
Bumped to driverkit v0.18.0 that contains needed change.
Dropping `wip` now.

File: cmd/driver/install/install.go [hunk 2]
```diff
@@ -119,12 +119,14 @@ func (o *driverInstallOptions) RunDriverInstall(ctx context.Context) (string, er
 		return "", nil
 	}
 
-	if o.Distro.String() == driverdistro.UndeterminedDistro && o.Compile {
-		o.Download = false
-		o.Printer.Logger.Info(
-			"Detected an unsupported target system, please get in touch with the Falco community. Trying to compile anyway.")
-	} else {
-		return "", fmt.Errorf("detected an unsupported target system, please get in touch with the Falco community")
+	if o.Distro.String() == driverdistro.UndeterminedDistro {
```
Comment (FedeDP) [inline cmd/driver/install/install.go:122]: We need to make the check only when distro is `undetermined`. This is a fix on top of current main (bug introduced in https://github.com/falcosecurity/falcoctl/pull/484)

File: pkg/driver/distro/distro.go [hunk 3]
```diff
@@ -179,6 +199,25 @@ func Build(ctx context.Context,
 	if err != nil {
 		return "", err
 	}
+
+	if env == nil {
```
Comment (FedeDP) [inline pkg/driver/distro/distro.go:203]: Avoid a segfault when later accessing the map.

File: pkg/driver/distro/distro.go [hunk 4]
```diff
@@ -179,6 +199,25 @@ func Build(ctx context.Context,
...
+		env = make(map[string]string)
+	}
+
+	// If customizeBuild did not set any KERNELDIR env variable,
+	// try to load kernel headers urls from driverkit.
+	if _, found := env[drivertype.KernelDirEnv]; !found {
```
Comment (FedeDP) [inline pkg/driver/distro/distro.go:209]: If no `customizeBuild` added the `KERNELDIR` env var to the env map, we try to autoamtically download and extract kernel headers leveraging driverkit; if we are able to do so, we emplace the new `KERNELDIR` env var pointing to the extraction path for the kernel headers.

File: pkg/driver/type/kmod.go [hunk 5]
```diff
@@ -61,26 +53,26 @@ func (k *kmod) String() string {
 // Then, using dkms, it tries to fetch all
 // dkms-installed versions of the module to clean them up.
 func (k *kmod) Cleanup(printer *output.Printer, driverName string) error {
-	_, err := exec.Command("bash", "-c", "hash lsmod").Output()
+	lsmod, err := exec.LookPath("lsmod")
```
Comment (FedeDP) [inline pkg/driver/type/kmod.go:56]: Switch to use `exec.LookPath` instead of the ugly bash magic.

File: pkg/driver/type/kmod.go [hunk 6]
```diff
@@ -223,8 +215,14 @@ func (k *kmod) Build(ctx context.Context,
 			printer.Logger.Warn("Could not fill /tmp/falco-dkms-make content.")
 			continue
 		}
-		dkmsCmdArgs := fmt.Sprintf(`dkms install --directive="MAKE='/tmp/falco-dkms-make'" -m %q -v %q -k %q --verbose`,
-			driverName, driverVersion, kr.String())
+		var dkmsCmdArgs string
+		if kernelDir, found := env[KernelDirEnv]; found {
+			dkmsCmdArgs = fmt.Sprintf(`dkms install --kernelsourcedir %q --directive="MAKE='/tmp/falco-dkms-make'" -m %q -v %q -k %q --verbose`,
```
Comment (FedeDP) [inline pkg/driver/type/kmod.go:220]: If `KERNELDIR` env var was passed, forward it to `dkms install` too!

Comment (FedeDP) [reply pkg/driver/type/kmod.go:220]: Note: `ebpf` did not need any change since it was already passing all the env to the `make` command.

**Review by leogr — APPROVED** [review]

**Pull request merged**